### PR TITLE
ARROW-10885: [Rust][DataFusion] Optimize hash join build vs probe order based on number of rows

### PR DESCRIPTION
**Pull request by Dandandan**
This PR uses the `num_rows` statistics to implement a common optimization to use the smallest table for the build phase.
This is a good heuristic, as to have the smallest table used in the build phase leads to less items to be inserted to the hash table, in particular if the size of tables is very imbalanced.

Some notes:

* The optimization works on the `LogicalPlan` by swapping left and right, the join type and the key order. This seems currently the easiest place to add it, as there is no cost based optimizer and/or optimizers on the physical plan yet. The optimization rule assumes that the left part of the join will be used for the build phase and the right part for the probe phase.
* It requires the number of rows to be exactly known, so it will not work whenever there is a transformation changing the number of rows, except for `limit`. The idea here is that in other cases, it is very hard to estimate the number of resulting rows.
* The impact currently is measurable on queries with a bigger left side of an (inner) join

 FYI @andygrove @jorgecarleitao

### CODE REVIEW

File: rust/datafusion/src/optimizer/hash_build_probe_order.rs
Comment (Dandandan) [inline rust/datafusion/src/optimizer/hash_build_probe_order.rs:43]: Not sure if this is relevant, where this is used?

File: rust/datafusion/src/optimizer/hash_build_probe_order.rs
Comment (andygrove) [inline rust/datafusion/src/optimizer/hash_build_probe_order.rs:39]: We should also cover `Join` here too? Since we only support inner joins for now, the output row count will at most be the number of rows of the smaller input.

File: rust/datafusion/src/optimizer/hash_build_probe_order.rs
Comment (andygrove) [inline rust/datafusion/src/optimizer/hash_build_probe_order.rs:94]: in theory, this is unnecessary since there are no restrictions in SQL on order of join conditions. However, it is possible we do make some assumptions so if you ean into that it would be good to file an issue for it.

Comment (Dandandan) [reply rust/datafusion/src/optimizer/hash_build_probe_order.rs:94]: Makes sense, was surprised by it that I had to. Currently it fails when you change the order (e.g. in a query already)

File: rust/datafusion/src/optimizer/hash_build_probe_order.rs
Comment (andygrove) [inline rust/datafusion/src/optimizer/hash_build_probe_order.rs:89]: I'm not sure if it is relevant, but when I have done this in other projects, I have wrapped the swapped join in a projection to preserve the column ordering of the output. This would be less surprising to a user.

**Comment by andygrove** [comment]
Thanks @Dandandan this is looking great :rocket:

File: rust/datafusion/src/optimizer/hash_build_probe_order.rs
Comment (Dandandan) [reply rust/datafusion/src/optimizer/hash_build_probe_order.rs:89]: Isn't the order explicit in the `schema` (which is the same) or will it be changed based on swapping left and right?

File: rust/datafusion/src/optimizer/hash_build_probe_order.rs
Comment (Dandandan) [reply rust/datafusion/src/optimizer/hash_build_probe_order.rs:39]: :+1:

File: rust/datafusion/src/optimizer/hash_build_probe_order.rs
Comment (andygrove) [reply rust/datafusion/src/optimizer/hash_build_probe_order.rs:89]: We should add a unit test for this so that we know the answer to that question, I think.

File: rust/datafusion/src/optimizer/hash_build_probe_order.rs
Comment (andygrove) [inline rust/datafusion/src/optimizer/hash_build_probe_order.rs:38]: I have been thinking about what we can do to estimate the number of rows coming out of joins so that we can extend this optimization to nested joins. We can't do anything accurate with the current statistics in this case but I feel that we should try and do something rather than just pick the left side as the build side.

One idea is to assume that all joins produce a cartesian product (left row count * right row count). This would at least help in the case where two small tables are joined, and then joined with a huge table, or the other way around.

File: rust/datafusion/src/optimizer/hash_build_probe_order.rs
Comment (andygrove) [reply rust/datafusion/src/optimizer/hash_build_probe_order.rs:43]: This looks good. This is used for projections without an input, such as `SELECT 1`.

File: rust/datafusion/src/optimizer/hash_build_probe_order.rs
Comment (Dandandan) [reply rust/datafusion/src/optimizer/hash_build_probe_order.rs:38]: Indeed I think that could be very beneficial but estimating it before executing can be really hard / impossible.

Also, if using the left as build side wrong, at this moment, the order could be changed by the user by changing the query itself.

I think ideally you should be able to know more about the table size when the query is executing (a la Spark 3 adaptive query execution) so you don't do the wrong thing. BigQuery also has a nice strategy / explanation for this https://cloud.google.com/bigquery/query-plan-explanation This probably requires quite a bit of changes on the execution / planning side, but this would bring much more available statistics to each step during execution to be able to change optimize the plan further.

Comment (andygrove) [reply rust/datafusion/src/optimizer/hash_build_probe_order.rs:38]: Some databases use the `STRAIGHT_JOIN` to force joins to happen in the user-specified order. This is from Impala docs:

```
If statistics are not available for all the tables in the join query, or if Impala chooses a join order that is not the most efficient, you can override the automatic join order optimization by specifying the STRAIGHT_JOIN keyword immediately after the SELECT and any DISTINCT or ALL keywords. In this case, Impala uses the order the tables appear in the query to guide how the joins are processed. 
```

I think we can merge this PR as is and continue this discussion. Spark's AQE approach would mean that we have the statistics, but only if we load both sides into memory first (or scan them first for row counts) which would possibly defeat the point of this optimization. It would also mean that the next operator in the query plan wouldn't be able to start streaming until the join has completed? This is a tricky area.

Comment (andygrove) [reply rust/datafusion/src/optimizer/hash_build_probe_order.rs:38]: I filed https://issues.apache.org/jira/browse/ARROW-10964 for "Optimize nested joins" and referenced this discussion.

**Review by andygrove — APPROVED** [review]

**Comment by andygrove** [comment]
I do think that we should start looking at optimizations on the physical plan and eventually move this optimization there. I also do think that an adaptive execution approach makes sense, especially in a distributed context. I think it might not work so well for the current single node / in-process execution approach though.

File: rust/datafusion/src/optimizer/hash_build_probe_order.rs
Comment (Dandandan) [reply rust/datafusion/src/optimizer/hash_build_probe_order.rs:98]: Filed here https://issues.apache.org/jira/browse/ARROW-10965

**Comment by seddonm1** [comment]
This is awesome work @Dandandan 👍

**Comment by Dandandan** [comment]
Also related to https://github.com/apache/arrow/pull/8965 which stops generating/using an index for the probe side.

**Comment by Dandandan** [comment]
I checked merging this https://github.com/apache/arrow/pull/8965 

Besides being ~20-50x faster regardless of this PR, reordering gives a further ~15% reduction in time when reordering the following query (6001214 vs 1499999 rows)

```
select
                l_shipmode,
                sum(case
                    when o_orderpriority = '1-URGENT'
                        or o_orderpriority = '2-HIGH'
                        then 1
                    else 0
                end) as high_line_count,
                sum(case
                    when o_orderpriority <> '1-URGENT'
                        and o_orderpriority <> '2-HIGH'
                        then 1
                    else 0
                end) as low_line_count
            from
                lineitem
            join
                orders
            on
                l_orderkey = o_orderkey
            group by
                l_shipmode
            order by
                l_shipmode;"
```

**Comment by Dandandan** [comment]
I wrote some details of the PRs for a planned blog post.

https://docs.google.com/document/d/1Urxm34rl8DZ5D0vyhlrrBoZK6IHW7WFRN3hsaTfPujg/edit?usp=drivesdk

**Comment by andygrove** [comment]
@Dandandan This needs rebasing - I tried merging into master locally before merging this and got some compilation errors.

```
error[E0050]: method `scan` has 3 parameters but the declaration in trait `datasource::datasource::TableProvider::scan` has 4
   --> datafusion/src/optimizer/hash_build_probe_order.rs:179:13
    |
179 | /             &self,
180 | |             _projection: &Option<Vec<usize>>,
181 | |             _batch_size: usize,
    | |______________________________^ expected 4 parameters, found 3
    | 
   ::: datafusion/src/datasource/datasource.rs:66:9
    |
66  | /         &self,
67  | |         projection: &Option<Vec<usize>>,
68  | |         batch_size: usize,
69  | |         filters: &[Expr],
    | |________________________- trait requires 4 parameters

error: cannot construct `plan::LogicalPlan` with struct literal syntax due to inaccessible fields
   --> datafusion/src/optimizer/hash_build_probe_order.rs:204:23
    |
204 |         let lp_left = LogicalPlan::TableScan {
    |                       ^^^^^^^^^^^^^^^^^^^^^^

error: cannot construct `plan::LogicalPlan` with struct literal syntax due to inaccessible fields
   --> datafusion/src/optimizer/hash_build_probe_order.rs:211:24
    |
211 |         let lp_right = LogicalPlan::TableScan {
    |                        ^^^^^^^^^^^^^^^^^^^^^^

```

**Comment by Dandandan** [comment]
@andygrove thanks, will do. Will also enable it now that #8965 is merged.

**Comment by Dandandan** [comment]
@andygrove updated & enabled the optimization now.

**Comment by andygrove** [comment]
I merged with master locally and tested it and see a speedup. I will merge when CI is green. Thanks @Dandandan

**Pull request closed without merging**
